syntax = "proto3";

package zen.remote;

import "common.proto";

service GlBaseTechniqueService
{
  rpc New(NewGlBaseTechniqueRequest) returns (EmptyResponse);

  rpc Delete(DeleteResourceRequest) returns (EmptyResponse);

  rpc BindProgram(BindProgramRequest) returns (EmptyResponse);

  rpc BindVertexArray(BindVertexArrayRequest) returns (EmptyResponse);

  rpc GlUniformVector(GlUniformVectorRequest) returns (EmptyResponse);

  rpc GlUniformMatrix(GlUniformMatrixRequest) returns (EmptyResponse);

  rpc GlDrawArrays(GlDrawArraysRequest) returns (EmptyResponse);
}

message NewGlBaseTechniqueRequest
{
  uint64 id = 1;
  uint64 rendering_unit_id = 2;
}

message BindProgramRequest
{
  uint64 id = 1;
  uint64 program_id = 2;
}

message BindVertexArrayRequest
{
  uint64 id = 1;
  uint64 vertex_array_id = 2;
}

message GlUniformVectorRequest
{
  uint64 id = 1;
  int32 location = 2;
  string name = 3;
  uint32 type = 4;  // enum UniformVariableType
  uint32 size = 5;
  uint32 count = 6;
  bytes value = 7;
}

message GlUniformMatrixRequest
{
  uint64 id = 1;
  int32 location = 2;
  string name = 3;
  uint32 col = 4;
  uint32 row = 5;
  uint32 count = 6;
  bool transpose = 7;
  bytes value = 8;
}

message GlDrawArraysRequest
{
  uint64 id = 1;
  uint32 mode = 2;
  int32 first = 3;
  uint32 count = 4;
}
